plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module UntypedPlutusCore.Check.Scope (checkScope) where
+
+import Control.Lens hiding (index)
+import UntypedPlutusCore.Core.Type as UPLC
+import UntypedPlutusCore.DeBruijn as UPLC
+
+import Control.Monad.Error.Lens
+import Control.Monad.Except
+
+{- | A pass to check that the input term:
+1) does not contain free variables and
+2) that all binders are set to debruijn index 0.
+
+Feeding the result of the debruijnification to this function is expected to pass.
+
+On the other hand, because of (2), this pass is
+stricter than the undebruijnification's (indirect)
+scope-checking, see NOTE: [DeBruijn indices of Binders].
+-}
+checkScope :: forall e m name uni fun a.
+             (HasIndex name, MonadError e m, AsFreeVariableError e)
```

Comment:
I'd maybe have had an error just for this pass, but reusing that one seems fine.